libut/Makefile

```diff
@@ -1,33 +1,41 @@
-OBJS=libut.a
-all: $(OBJS) 
+SHELL=/usr/sh
+VERSION=2.0.1
+MAJOR=2
+SOURCES=$(shell echo src/*.c)
+HEADERS=$(shell echo include/*.h)
+OBJS=$(SOURCES:.c=.o)
+all: libut.a libut.so.$(VERSION)
 INCDIR=./include
+LIBDIR=/usr/lib
 CFLAGS+=-I$(INCDIR)
-CFLAGS+=-Wall -Wextra
+CFLAGS+=-Wall -Wextra -Werror -fPIC
 CFLAGS+=-g
+CFLAGS+=$(OPTFLAGS)
 
-libut.a: libut.o utvector.o utmm.o ringbuf.o
+libut.a: $(OBJS)
 	ar r $@ $^
 
-libut.o: src/libut.c $(INCDIR)/libut.h
-	$(CC) $(CFLAGS) $(CPPFLAGS) -c $<
-
-utvector.o: src/utvector.c $(INCDIR)/utvector.h
-	$(CC) $(CFLAGS) $(CPPFLAGS) -c $<
-
-utmm.o: src/utmm.c $(INCDIR)/utmm.h
-	$(CC) $(CFLAGS) $(CPPFLAGS) -c $<
-
-ringbuf.o: src/ringbuf.c $(INCDIR)/ringbuf.h
-	$(CC) $(CFLAGS) $(CPPFLAGS) -c $<
+libut.so.$(VERSION): $(OBJS)
+	$(CC) $(CFLAGS) -shared -Wl,-soname,libut.so.$(MAJOR) $^ -o $@
 
 .PHONY: clean tests install
 
 clean:
-	rm -f $(OBJS) *.o
+	rm -f *.o *.a *.so*
 	make -C tests clean
 
 tests: libut.a
 	make -C tests 
 
-install: libut.a
-	cp $< /usr/local/lib
+install: libut.a libut.so.$(VERSION)
+	mkdir -p $(DESTDIR)$(LIBDIR)
+	mkdir -p $(DESTDIR)$(LIBDIR)/pkgconfig
+	mkdir -p $(DESTDIR)/usr/include/uthash
+	cp libut.a $(DESTDIR)$(LIBDIR)
+	cp libut.so.$(VERSION) $(DESTDIR)$(LIBDIR)
+	ln -sf $(LIBDIR)/libut.so.$(VERSION) $(DESTDIR)$(LIBDIR)/libut.so.$(MAJOR)
+	ln -sf $(LIBDIR)/libut.so.$(VERSION) $(DESTDIR)$(LIBDIR)/libut.so
+	cp ./include/*.h $(DESTDIR)/usr/include/uthash
```

Comment:
> extraordinary evidence

> since we have many ut\*.h header files, not just one or two. then ***why not*** gather them into a directory/namespace and leave an universal header outside in /usr/include? that ***looks just nice***...

No, see, that's not an extraordinarily convincing argument. That's no argument at all. You *literally* used the phrase "why not?". The reason "not" is clear: inertia, status quo, don't fix what ain't broke, be very careful not to break any existing users, Occam's Razor, KISS, etc. I'm looking for you to provide a positive reason "for" changing the current system.

I looked for "opensuse packages depending on uthash" and found https://build.opensuse.org/package/show/home:layus/pocl : it uses the header `utlist.h` but not `libut.*`, and it currently vendors `utlist.h` into its own source tree instead of relying on the SUSE package anyway.

I also found https://build.opensuse.org/package/show/Archiving/archivemount , which uses `uthash.h` but not `libut.*`, and vendors `uthash.h` into its own source tree.

I'm fairly confident that OpenSUSE should just stop trying to distribute libut.a to end-users (i.e. follow Debian's lead here), and then all the problems vanish. If you want to put the header files in /usr/include/uthash, that seems fine to me, but I feel like it would break Debian to take that patch upstream.